### PR TITLE
Release 4.1.0

### PR DESCRIPTION
**Pull request by bjcfuller**
## What's new in 4.1.0

CL 4.1.0 is a minor release focused on code quality and adding new components.

* Adds a Promo component for displaying timely promotional information (shortcode support only) (#177)
* Adds a Date component for displaying dates (shortcode support only) (#176)
  - Generates a downloadable ICS file
  - Downloaded state persists for 12 hours
* All scripts are now linted with ESLint; JSCS and JSHint have been removed (#179)
* Panels formatted as `super` now have a bottom margin by default
* Fixes an issue that prevented CSS classes added in the Block Editor Inspector from being passed to components
* Fixes an issue with padding on cards placed in equal-height tiles
* Updates development tools

### CODE REVIEW

File: src/js/dates.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+		const c = el.querySelector( '.cl-date-download-cancel' );
+		c.addEventListener( 'click', dialogue.bind( null, el ), false );
+
+		const hash = el.getAttribute( 'data-hash' );
+		const cvalue = URICL.getCookie( 'uri-cl-' + hash );
+
+		if ( 'downloaded' === cvalue ) {
+			el.classList.add( 'downloaded' );
+			p.setAttribute( 'title', "You've downloaded this date before. Check your downloads folder for an ICS file." );
```
Comment (bjcfuller) [inline src/js/dates.js]: Codacy Issue found: [Strings must use singlequote.](https://app.codacy.com/manual/uriweb/uri-component-library/pullRequest?prid=4835245)

**Comment by bjcfuller** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
- Added 7
           

Complexity increasing per file
==============================
- src/js/dates.js  3
         

Complexity decreasing per file
==============================
+ src/js/dismiss.js  -1
         

Clones added
============
- src/js/blocks/card/block.js  2
- src/js/blocks/hero/block.js  1
         

Clones removed
==============
+ src/js/blocks/button/block.js  -1
+ src/js/blocks/notice/block.js  -1
+ src/js/blocks/panel/block.js  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/uriweb/uri-component-library/pullRequest?prid=4835245&bid=16060032)

File: src/js/dates.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	function setupDate( el ) {
+		el.classList.add( 'has-js' );
+		const p = el.querySelector( '.cl-date-content-wrapper' );
+		p.addEventListener( 'click', dialogue.bind( null, el ), false );
+
+		const a = el.querySelector( '.cl-date-download-confirm' );
+		a.addEventListener( 'click', download.bind( null, el, p ), false );
+
+		const c = el.querySelector( '.cl-date-download-cancel' );
+		c.addEventListener( 'click', dialogue.bind( null, el ), false );
```
Comment (bjcfuller) [inline src/js/dates.js]: Codacy Issue found: ['dialogue' was used before it was defined.](https://app.codacy.com/manual/uriweb/uri-component-library/pullRequest?prid=4835245)

File: src/js/dates.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+		}
+	}
+
+	function setupDate( el ) {
+		el.classList.add( 'has-js' );
+		const p = el.querySelector( '.cl-date-content-wrapper' );
+		p.addEventListener( 'click', dialogue.bind( null, el ), false );
+
+		const a = el.querySelector( '.cl-date-download-confirm' );
+		a.addEventListener( 'click', download.bind( null, el, p ), false );
```
Comment (bjcfuller) [inline src/js/dates.js]: Codacy Issue found: ['download' was used before it was defined.](https://app.codacy.com/manual/uriweb/uri-component-library/pullRequest?prid=4835245)

File: src/js/dates.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+		for ( i = 0; i < els.length; i++ ) {
+			setupDate( els[ i ] );
+		}
+	}
+
+	function setupDate( el ) {
+		el.classList.add( 'has-js' );
+		const p = el.querySelector( '.cl-date-content-wrapper' );
+		p.addEventListener( 'click', dialogue.bind( null, el ), false );
```
Comment (bjcfuller) [inline src/js/dates.js]: Codacy Issue found: ['dialogue' was used before it was defined.](https://app.codacy.com/manual/uriweb/uri-component-library/pullRequest?prid=4835245)

File: src/js/dates.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
+/**
+ * DATES
+ *
+ * @package uri-component-library
+ */
+
+( function() {
+	'use strict';
+
+	window.addEventListener( 'DOMContentLoaded', init, false );
```
Comment (bjcfuller) [inline src/js/dates.js]: Codacy Issue found: ['init' was used before it was defined.](https://app.codacy.com/manual/uriweb/uri-component-library/pullRequest?prid=4835245)

File: src/js/dates.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+	function init() {
+		let i;
+
+		const els = document.querySelectorAll( '.cl-date' );
+
+		for ( i = 0; i < els.length; i++ ) {
+			setupDate( els[ i ] );
```
Comment (bjcfuller) [inline src/js/dates.js]: Codacy Issue found: ['setupDate' was used before it was defined.](https://app.codacy.com/manual/uriweb/uri-component-library/pullRequest?prid=4835245)

Comment (bjcfuller) [inline src/js/dates.js]: Codacy Issue found: [Function Call Object Injection Sink](https://app.codacy.com/manual/uriweb/uri-component-library/pullRequest?prid=4835245)

**Comment by bjcfuller** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
- Added 7
           

Complexity increasing per file
==============================
- src/js/dates.js  3
         

Complexity decreasing per file
==============================
+ src/js/dismiss.js  -1
         

Clones added
============
- src/js/blocks/card/block.js  2
- src/js/blocks/hero/block.js  1
         

Clones removed
==============
+ src/js/blocks/panel/block.js  -1
+ src/js/blocks/notice/block.js  -1
+ src/js/blocks/button/block.js  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/uriweb/uri-component-library/pullRequest?prid=4835245&bid=16060032)

File: src/js/dates.js (hunk 5 above)
Comment (bjcfuller) [inline src/js/dates.js]: Codacy Issue found: ['init' was used before it was defined.](https://app.codacy.com/manual/uriweb/uri-component-library/pullRequest?prid=4835245)

File: src/js/dates.js (hunk 2 above)
Comment (bjcfuller) [inline src/js/dates.js]: Codacy Issue found: ['dialogue' was used before it was defined.](https://app.codacy.com/manual/uriweb/uri-component-library/pullRequest?prid=4835245)

File: src/js/dates.js (hunk 3 above)
Comment (bjcfuller) [inline src/js/dates.js]: Codacy Issue found: ['download' was used before it was defined.](https://app.codacy.com/manual/uriweb/uri-component-library/pullRequest?prid=4835245)

File: src/js/dates.js (hunk 6 above)
Comment (bjcfuller) [inline src/js/dates.js]: Codacy Issue found: [Function Call Object Injection Sink](https://app.codacy.com/manual/uriweb/uri-component-library/pullRequest?prid=4835245)

File: src/js/dates.js (hunk 4 above)
Comment (bjcfuller) [inline src/js/dates.js]: Codacy Issue found: ['dialogue' was used before it was defined.](https://app.codacy.com/manual/uriweb/uri-component-library/pullRequest?prid=4835245)

File: src/js/dates.js (hunk 6 above)
Comment (bjcfuller) [inline src/js/dates.js]: Codacy Issue found: ['setupDate' was used before it was defined.](https://app.codacy.com/manual/uriweb/uri-component-library/pullRequest?prid=4835245)

File: src/js/dates.js (hunk 1 above)
Comment (bjcfuller) [inline src/js/dates.js]: Codacy Issue found: [Strings must use singlequote.](https://app.codacy.com/manual/uriweb/uri-component-library/pullRequest?prid=4835245)

**Comment by johnpennypacker** [comment]
Solid!

**Comment by bjcfuller** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
- Added 7
           

Complexity increasing per file
==============================
- src/js/dates.js  3
         

Complexity decreasing per file
==============================
+ src/js/dismiss.js  -1
         

Clones added
============
- src/js/blocks/card/block.js  2
- src/js/blocks/hero/block.js  1
         

Clones removed
==============
+ src/js/blocks/notice/block.js  -1
+ src/js/blocks/panel/block.js  -1
+ src/js/blocks/button/block.js  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/uriweb/uri-component-library/pullRequest?prid=4835245&bid=16060032)

File: src/js/dates.js (hunk 4 above)
Comment (bjcfuller) [inline src/js/dates.js]: Codacy Issue found: ['dialogue' was used before it was defined.](https://app.codacy.com/manual/uriweb/uri-component-library/pullRequest?prid=4835245)

File: src/js/dates.js (hunk 6 above)
Comment (bjcfuller) [inline src/js/dates.js]: Codacy Issue found: ['setupDate' was used before it was defined.](https://app.codacy.com/manual/uriweb/uri-component-library/pullRequest?prid=4835245)

File: src/js/dates.js (hunk 2 above)
Comment (bjcfuller) [inline src/js/dates.js]: Codacy Issue found: ['dialogue' was used before it was defined.](https://app.codacy.com/manual/uriweb/uri-component-library/pullRequest?prid=4835245)

File: src/js/dates.js (hunk 3 above)
Comment (bjcfuller) [inline src/js/dates.js]: Codacy Issue found: ['download' was used before it was defined.](https://app.codacy.com/manual/uriweb/uri-component-library/pullRequest?prid=4835245)

File: src/js/dates.js (hunk 1 above)
Comment (bjcfuller) [inline src/js/dates.js]: Codacy Issue found: [Strings must use singlequote.](https://app.codacy.com/manual/uriweb/uri-component-library/pullRequest?prid=4835245)

File: src/js/dates.js (hunk 6 above)
Comment (bjcfuller) [inline src/js/dates.js]: Codacy Issue found: [Function Call Object Injection Sink](https://app.codacy.com/manual/uriweb/uri-component-library/pullRequest?prid=4835245)

File: src/js/dates.js (hunk 5 above)
Comment (bjcfuller) [inline src/js/dates.js]: Codacy Issue found: ['init' was used before it was defined.](https://app.codacy.com/manual/uriweb/uri-component-library/pullRequest?prid=4835245)

**Pull request merged**